src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AvoidDuplicateElementInitialization.cs

```diff
@@ -0,0 +1,142 @@
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftQualityGuidelinesAnalyzersResources.AvoidDuplicateElementInitializationTitle), MicrosoftQualityGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftQualityGuidelinesAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftQualityGuidelinesAnalyzersResources.AvoidDuplicateElementInitializationMessage), MicrosoftQualityGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftQualityGuidelinesAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftQualityGuidelinesAnalyzersResources.AvoidDuplicateElementInitializationDescription), MicrosoftQualityGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftQualityGuidelinesAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessage,
+                                                                             DiagnosticCategory.Usage,
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: true,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        private static readonly SyntaxGenerator s_generator = SyntaxGenerator.GetGenerator(new AdhocWorkspace(), LanguageNames.CSharp);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationAction(AnalyzeOperation, OperationKind.ObjectCreation);
+        }
+
+        private static void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var objectInitializer = (IObjectCreationOperation)context.Operation;
+            if (objectInitializer.Initializer == null)
+            {
+                return;
+            }
+
+            var initializedElementIndexes = new HashSet<object[]>(ConstantArgumentEqualityComparer.Instance);
```

Comment:
Can we use `SequenceEquals` and compare each element with `ConstantValue.Equals`? Are there cases where this fails and you need a custom comparer?

Reply:
I'm not quite sure I follow. I updated the custom comparer to use `SequenceEqual` since that is cleaner (thank you), but I'm not sure how it would let me avoid a custom comparer altogether. Could you elaborate please?

Reply:
I meant can we use the default equality comparer, which will mean that the equality check will invoke this virtual method: http://source.roslyn.io/#Microsoft.CodeAnalysis/ConstantValue.cs,733. For constants with int32 underlying value, it will end up here: http://source.roslyn.io/#Microsoft.CodeAnalysis/ConstantValueSpecialized.cs,670, which should return true if they are of same type and have the same int32 value.

Reply:
Well I see that I can change this to a `HashSet<ConstantValue[]>` and then in the custom comparer, compare `ConstantValue` instances instead of `object` instance (although I'm not sure what additional benefit that has). But how would I compare two arrays of `ConstantValue` without using a custom comparer? I may be missing something. Could you maybe show me how you would rewrite this line?

Reply:
Ah, I forgot that default equality comparer for arrays will check for reference equality, not compare contents of array. Your approach seems fine.

Reply:
Ok thanks. I'll push the latest updates then.